src/syntaxtree.c

```diff
@@ -0,0 +1,898 @@
+        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+    }
+
+#define COMPILER_(tnum, ...)                                                 \
+    COMPILER(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+
+#define ARG(name, func) \
+    { name, func }
+#define ARG_(name) ARG(name, SyntaxTreeCompiler)
+
+static Obj SyntaxTreeFunc(Obj result, Obj func);
+
+static inline Obj NewSyntaxTreeNode(const char * type)
+{
+    Obj result;
+    Obj typestr;
+
+    C_NEW_STRING_DYN(typestr, type);
+    result = NEW_PREC(1);
+    AssPRec(result, RNamName("type"), typestr);
+
+    return result;
+}
+
+static Obj SyntaxTreeCompiler(Expr expr)
+{
+    Obj       result;
+    UInt      tnum;
+    CompilerT comp;
+
+    // TODO: GAP_ASSERT
+    tnum = TNUM_EXPR(expr);
+    comp = Compilers[tnum];
+
+    // TODO: clean
+    result = NewSyntaxTreeNode(comp.name);
+    // AssPRec(result, RNamName("line"), INTOBJ_INT(STAT_HEADER(expr)->line));
+
+    comp.compile(result, expr);
+
+    return result;
+}
+
+static Obj SyntaxTreeIntObjInt(UInt i)
+{
+    return INTOBJ_INT(i);
+}
+
+static Obj SyntaxTreeDefaultCompiler(Obj result, Expr expr)
+{
+    int       i;
+    UInt      tnum;
+    CompilerT comp;
+
+    // TODO: GAP_ASSERT tnum range
+    tnum = TNUM_EXPR(expr);
+    comp = Compilers[tnum];
+
+    for (i = 0; i < comp.arity; i++) {
+        AssPRec(result, RNamName(comp.args[i].argname),
+                comp.args[i].argcomp(ADDR_EXPR(expr)[i]));
+    }
+    return result;
+}
+
+Obj SyntaxTreeFunccall(Obj result, Expr expr)
+{
+    Obj  func;
+    Obj  args, argi;
+    UInt narg, i;
+
+    func = SyntaxTreeCompiler(FUNC_CALL(expr));
+    AssPRec(result, RNamName("funcref"), func);
+
+    /* compile the argument expressions */
+    narg = NARG_SIZE_CALL(SIZE_EXPR(expr));
+    args = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(args, narg);
+
+    for (i = 1; i <= narg; i++) {
+        argi = SyntaxTreeCompiler(ARGI_CALL(expr, i));
+        SET_ELM_PLIST(args, i, argi);
+        CHANGED_BAG(args);
+    }
+    AssPRec(result, RNamName("args"), args);
+    return result;
+}
+
+Obj SyntaxTreeFuncExpr(Obj result, Expr expr)
+{
+    Obj fexs;
+    Obj fexp;
+
+    fexs = FEXS_FUNC(CURR_FUNC());
+    fexp = ELM_PLIST(fexs, ((Int *)ADDR_EXPR(expr))[0]);
+
+    SyntaxTreeFunc(result, fexp);
+
+    return result;
+}
+
+Obj SyntaxTreeIntExpr(Obj result, Expr expr)
+{
+    Obj  value;
+    UInt size;
+    UInt typ;
+
+    if (IS_INTEXPR(expr))
+        value = OBJ_INTEXPR(expr);
+    else {
+        size = SIZE_EXPR(expr) - sizeof(UInt);
+        typ = *(UInt *)ADDR_EXPR(expr);
+        value = C_MAKE_INTEGER_BAG(size, typ);
+        memcpy(ADDR_OBJ(value), (void *)(&ADDR_EXPR(expr)[1]), size);
+    }
+
+    AssPRec(result, RNamName("value"), value);
+
+    return result;
+}
+
+Obj SyntaxTreeCharExpr(Obj result, Expr expr)
+{
+    Obj value;
+
+    value = NewBag(T_CHAR, 1);
+    *(UChar*)ADDR_OBJ(value) = (UChar)ADDR_EXPR(expr)[0];
```

Comment:
Those two lines are wrong, I am afraid -- T_CHAR instances are singletons, so you should use `ObjsChar[i]` instead; see also the code of `EvalCharExpr`.